File: source/Iron/Rendering/Sprite.cpp
#include "Sprite.h"

void Sprite::SetAsSpriteSheet(int tileWidth, int tileHeight)
{
    IsSpriteSheet = true;
    TileWidth = tileWidth;
    TileHeight = tileHeight;
}

void Sprite::GetTexCoord(int tileIndex, glm::vec2* texCoords)
{
    float tw = float(TileWidth) / float(Width);
    float th = float(TileHeight) / float(Height);
    int numPerRow = Width / TileWidth;
    float tx = (tileIndex % numPerRow) * tw;
    float ty = (tileIndex / numPerRow) * th;

    texCoords[0] = glm::vec2(tx + tw, ty);
    texCoords[1] = glm::vec2(tx + tw, ty + th);
    texCoords[2] = glm::vec2(tx, ty);
    texCoords[3] = glm::vec2(tx, ty + th);
}

int Sprite::TilesCount()
{
    return IsSpriteSheet ? Width / TileWidth * Height / TileHeight : 1;
}

File: api/IronCore/Rendering/Sprite.cs
using System.Runtime.CompilerServices;

namespace Iron
{
    public class Sprite
    {
        internal Sprite(uint id)
        {
            ID = id;
        }
        
        public uint ID { get; private set; }

        public int Width => GetWidth_Internal(ID);
        public int Height => GetHeight_Internal(ID);

        public void SetAsSpriteSheet(int tileWidth, int tileHeight)
        {
            SetAsSpriteSheet_Internal(ID, tileWidth, tileHeight);
        }

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void SetAsSpriteSheet_Internal(uint spriteID, int tileWidth, int tileHeight);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern int GetWidth_Internal(uint spriteID);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern int GetHeight_Internal(uint spriteID);
    }
}